test/integration/framework/master_utils.go

```diff
@@ -0,0 +1,230 @@
+	// Channel for stop signals to all master components
+	StopCh chan struct{}
+}
+
+type Config struct {
+	// If nil, a default is used, partially filled configs will not get populated.
+	MasterConfig            *master.Config
+	StartReplicationManager bool
+	// TODO: Add configs for endpoints controller, scheduler etc
+}
+
+func NewMasterComponents(c *Config) *MasterComponents {
+	m, s := startMasterOrDie(c.MasterConfig)
+	glog.Infof("Master %+v", s.URL)
+	DeleteAllEtcdKeys()
+	restClient := client.NewOrDie(&client.Config{Host: s.URL, Version: "v1beta3"})
+	stopCh := make(chan struct{})
+	controllerManager := controller.NewReplicationManager(restClient)
+	if c.StartReplicationManager {
+		controllerManager.RunUntil(1*time.Second, stopCh)
+	}
+	return &MasterComponents{
+		ApiServer:         s,
+		KubeMaster:        m,
+		RestClient:        restClient,
+		ControllerManager: controllerManager,
+		StopCh:            stopCh,
+	}
+}
+
+// startMasterOrDie starts a kubernetes master and an httpserver to handle api requests
+func startMasterOrDie(masterConfig *master.Config) (*master.Master, *httptest.Server) {
+	var m *master.Master
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		m.Handler.ServeHTTP(w, req)
+	}))
+	if masterConfig == nil {
+		helper, err := master.NewEtcdHelper(NewEtcdClient(), "")
+		if err != nil {
+			panic(err)
+		}
+		masterConfig = &master.Config{
+			EtcdHelper:        helper,
+			KubeletClient:     client.FakeKubeletClient{},
+			EnableLogsSupport: false,
+			EnableProfiling:   true,
+			EnableUISupport:   false,
+			APIPrefix:         "/api",
+			Authorizer:        apiserver.NewAlwaysAllowAuthorizer(),
+			AdmissionControl:  admit.NewAlwaysAdmit(),
+			SkipServices:      true,
+		}
+	}
+	m = master.New(masterConfig)
+	return m, s
+}
+
+func (m *MasterComponents) Stop() {
+	glog.Infof("Stopping master components")
+	// Ordering matters because the apiServer will only shutdown when pending
+	// requests are done
+	close(m.StopCh)
+	m.ApiServer.Close()
+}
+
+// Crate a selector with given field description.
+func CreateSelectorOrDie(fieldSelector string) fields.Selector {
+	field, err := fields.ParseSelector(fieldSelector)
+	if err != nil {
+		panic(err)
+	}
+	return field
+}
+
+type ClientUtils struct {
+	Client *client.Client
+	// Channel to shutdown all resources, eg: informers
+	StopCh <-chan struct{}
+	// Wait group used to sync reflector/client update operations
+	UpdateWg *sync.WaitGroup
+}
+
+// WatchPods creates an informer to watch pods matching given field selector.
+func (c *ClientUtils) Watch(field fields.Selector, resource, ns string) cache.Store {
+	lw := cache.NewListWatchFromClient(c.Client, resource, ns, field)
+	store, controller := fw.NewInformer(
+		lw, &api.Pod{}, 0, fw.ResourceEventHandlerFuncs{
+			UpdateFunc: func(obj, newObj interface{}) {
+				// Incremented as updates are sent by the client, waited on by the test
+				c.UpdateWg.Done()
```

Comment:
It's hard to see how this is going to work-- updates could be collapsed together (e.g., several updates appear to happen at once because a re-list happens).